### PR TITLE
Core: Support case sensitivity in MergingSnapshotProducer

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds support for case sensitivity in `MergingSnapshotProducer`.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
cc @jackye1995 @rdblue @RussellSpitzer @flyrain @karuppayya @szehon-ho @kbendick

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 1]
```diff
@@ -125,6 +127,16 @@ public ThisT set(String property, String value) {
     return self();
   }
 
+  protected void caseSensitiveBinding(boolean newCaseSensitive) {
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:130]: Other method argument is 'isCaseSensitive', should we follow?

Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:130]: Also wondering, method clashing is why this method can not be just caseSensitive(isCaseSensitive)?

**Review by szehon-ho — COMMENTED** [review]
Mostly looks good to me, very minor questions

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 1 above)
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:130]: Maybe we could be consistent with the name. I saw three names(caseSensitive, isCaseSensitive, newCaseSensitive) used for the parameters. Kind of prefer one simple name for all of them, like `caseSensitive`.

File: core/src/test/java/org/apache/iceberg/TestDeleteFiles.java [hunk 2]
```diff
@@ -268,6 +268,40 @@ public void testCannotDeleteFileWhereNotAllRowsMatchPartitionFilter() {
...
 
+  @Test
+  public void testDeleteCaseSensitivity() {
+    table.newFastAppend()
+        .appendFile(DATA_FILE_BUCKET_0_IDS_0_2)
+        .commit();
+
+    Expression rowFilter = Expressions.lessThan("iD", 5);
+
+    AssertHelpers.assertThrows("Should use case sensitive binding by default",
```
Comment (flyrain) [inline core/src/test/java/org/apache/iceberg/TestDeleteFiles.java:279]: Just curious, does it make sense to make non case sensitive the default?

**Review by flyrain — APPROVED** [review]
LGTM

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 1 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:130]: I think the argument name matters less than the method name. Why does this use `caseSensitiveBinding`? Why not add `caseSensitive` and use that as the implementation? We already have a type parameter for the class (`ThisT`) and a `self()` method so we can implement a chainable call here.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 3]
```diff
@@ -125,6 +127,16 @@ public ThisT set(String property, String value) {
...
+    this.caseSensitive = newCaseSensitive;
+    filterManager.caseSensitive(newCaseSensitive);
+    deleteFilterManager.caseSensitive(newCaseSensitive);
+  }
+
+  protected boolean caseSensitive() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:136]: I think this should be `isCaseSensitive`. I think we mostly use the `is` prefix for returning booleans like this.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 4]
```diff
@@ -271,7 +282,7 @@ protected void validateAddedDataFiles(TableMetadata base, Long startingSnapshotI
     Set<Long> newSnapshots = history.second();
 
     ManifestGroup conflictGroup = new ManifestGroup(ops.io(), manifests, ImmutableList.of())
-        .caseSensitive(caseSensitive)
+        .caseSensitive(caseSensitive())
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:285]: Nit: why access an instance field through the getter method?

File: core/src/test/java/org/apache/iceberg/TestDeleteFiles.java (hunk 2 above)
Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/TestDeleteFiles.java:279]: Case sensitive is more strict, so I think that makes sense as the default.

Comment (aokolnychyi) [reply core/src/test/java/org/apache/iceberg/TestDeleteFiles.java:279]: Iceberg historically uses case sensitivity resolution by default in all other places.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 1 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java]: Updated the var and method names.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 3 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java]: Fixed.

**Comment by rdblue** [comment]
Thanks, @aokolnychyi!

**Pull request merged**